### PR TITLE
Measure-Based Fixed Scaling

### PR DESCRIPTION
**Pull request by curran**
Will close **Measure-based fixed scaling** #12

### CODE REVIEW

File: js/dev/baton.js [hunk 1]
```diff
@@ -31,21 +31,28 @@ function parseJSON(proll) {
...
             time += 1 / numBeats;
         }
     });
-    console.log(measuresToBeats);
+
+    // Replace the starttime values (in beats) with our computed absolute time values.
+    proll.partdata.forEach(function (part){
+        part.notedata.forEach(function (d){
+            d.starttime[0] = beatsToTime[d.starttime[0]];
+            //d.duration[0] = d.duration[0] / beatsToTimeSignature;
```
Comment (seemantk) [inline js/dev/baton.js]: @curran, here I think is where we can add support for alternating.  If you were to add a new field, e.g.:
` d.beattime = beatsToTime[d.starttime[0]];`

then the original stuff remains intact, and the work becomes switching the scales (and dependencies) to use one or the other field.

thoughts?

Comment (curran) [reply js/dev/baton.js]: @seemantk :+1:  Yes, totally. Then a checkbox can be added to toggle between using `d.starttime[0]` and `d.beattime`.

**Comment by curran** [comment]
@seemantk I'm kind of hitting a wall with this, and I'd appreciate your input as to how to best proceed. I've got the values we need in `d.starttimeAbsolute` and `d.durationAbsolute`, where `d` is a single note. These fields are appended to the data elements inside `measureScaling.js`.

I've also wired up a UI checkbox for toggling the scaling.

What needs to be done when we hit the checkbox is:

 * Update the X scale domain to be based on `d.starttimeAbsolute` and `d.durationAbsolute`
 * Use the values for `d.starttimeAbsolute` and `d.durationAbsolute` in all places where currently we use `d.starttime[0]` and `d.duration[0]`.

Currently these are quite scattered, and I'm not sure how to best approach it.

One idea is to replace all occurrences of `d.starttime[0]` and `d.duration[0]` in the code with accessors `start(d)` and `duration(d)`. That way the logic for actually getting the value can be swapped out when the scaling mode is toggled. Then it would be a matter of figuring out what bits of the code need to re-run in order to update the rendered notes.

One additional complexity is that toggling the scaling also needs to apply to barlines.

**Comment by curran** [comment]
@seemantk Just a random thought - this is *exactly* the kind of situation where using an architecture like Redux + d3-component would be a huuuge win. The change would be trivial if everything were set up like that - centralized state, pure functional rendering logic. Food for thought :smile: (for future projects).

**Comment by seemantk** [comment]
@curran, sorry for the delay in looking at this.  I'll dive into this during this week and see how I can help.

**Comment by curran** [comment]
@seemantk I'm running into an issue with the note durations, maybe you would have some ideas on how to solve them.

The thing is, I need to use the measure-scaled duration in place of `d.duration[0]` here:

```js
          .attr("xlink:href", function(d) {
              return "#note-" + d.duration[0];
            })
```

The issue is that the scaled durations will not be whole numbers, they are values like 0.5, 0.375, 0.125, 0.25, etc, so we may need to change the SVG scheme.

**Comment by curran** [comment]
@seemantk Just had an insight that we can dynamically generate these based on the values that are present. That would totally work. Was worried before that we'need to foresee every possible fractional value.

File: js/dev/baton.js [hunk 2]
```diff
@@ -3,57 +3,58 @@ var margin = { top: 20, right: 20, bottom: 20, left: 20 }
...
+          opt.enter()
+            .append("option")
+              .property("value", function(d) { return d; })
+          ;
+        })
+  ;
+  /*
+  **  Activate the "Load button"
+  */
+  d3.select("#load_song").on("click", function() {
```
Comment (seemantk) [inline js/dev/baton.js]: This spacing might have been accidental, as the only change I see is the addition of the new signal here.  Unless I'm missing something?

File: js/dev/baton.js [hunk 3]
```diff
@@ -84,6 +85,8 @@ function load_song(work) {
             if(error) throw error;
             // Set the URL history to the current song
             history.pushState(null, null, '?id=' + work);
+
+            // Parse the raw JSON and pass it to chartify.
```
Comment (seemantk) [inline js/dev/baton.js:98]: 👍

File: js/dev/baton.js [hunk 4]
```diff
@@ -105,23 +108,26 @@ function setupDispatcher() {
...
-        .on("show-ribbon",  notesBook.ribbons)
-        .on("separate-voices", notesBook.separate)
-        .on("zoom",     notesBook.zoom)
+        .on("show-notes",            notesBook.notes)
+        .on("show-extremes",         notesBook.extremes)
+        .on("hilite",                notesBook.hilite)
+        .on("show-ribbon",           notesBook.ribbons)
+        .on("separate-voices",       notesBook.separate)
+        .on("zoom",                  notesBook.zoom)
+        .on("measure-based-scaling", notesBook.measureScaling)
```
Comment (seemantk) [inline js/dev/baton.js:126]: 👍  on the spacing here

**Comment by curran** [comment]
@seemantk Your changes in https://github.com/sul-cidr/josquin-ribbon/pull/138 unblocked this work woohoo! Now I can set the width of rects to a fractional value without issue.

This is coming along. The scaling for the notes is done, now we need to apply that same scaling to everything else, like bar lines and mensuration symbols. Probably we'll need to also recompute the X domain when this is toggled.

image

**Comment by curran** [comment]
Finally breaking through with this.

 * [x] Transformed note start times
 * [x] Transformed note durations
 * [x] Transformed barlines
 * [x] Transformed mensuration symbols
 * [x] Transformed section labels
 * [ ] Adjust the domain when toggling scaling
 * [x] Incorporate the changes from **Externalizing mensuration lookup** #140

image

image

**Comment by curran** [comment]
@seemantk This is ready for review.

I feel this work is presentable (perhaps ready for merge), but also has the following issues I'm aware of (that could possibly be addressed as separate issues/PRs):

 * The domain does not adjust.
 * The zoom brush state is not taken into account when toggling the measure scaling.

**Comment by curran** [comment]
Hey @seemantk , I just incorporated the changes from **Externalizing mensuration lookup** #140.

I'd like to propose we merge this as-is for discussion/feedback purposes, and address the remaining issues separately. What do you think?

**Pull request merged**